src/dmd/optimize.d

```diff
@@ -188,6 +188,52 @@ extern (C++) Expression fromConstInitializer(int result, Expression e1)
     return e;
 }
 
+/* It is possible for constant folding to change an array expression of
+ * unknown length, into one where the length is known.
+ * If the expression 'arr' is a literal, set lengthVar to be its length.
+ */
+package void setLengthVarIfKnown(VarDeclaration lengthVar, Expression arr)
+{
+    if (!lengthVar)
+        return;
+    if (lengthVar._init && !lengthVar._init.isVoidInitializer())
+        return; // we have previously calculated the length
+    size_t len;
+    if (arr.op == TOKstring)
+        len = (cast(StringExp)arr).len;
+    else if (arr.op == TOKarrayliteral)
+        len = (cast(ArrayLiteralExp)arr).elements.dim;
+    else
+    {
+        Type t = arr.type.toBasetype();
+        if (t.ty == Tsarray)
+            len = cast(size_t)(cast(TypeSArray)t).dim.toInteger();
+        else
+            return; // we don't know the length yet
+    }
+    Expression dollar = new IntegerExp(Loc(), len, Type.tsize_t);
+    lengthVar._init = new ExpInitializer(Loc(), dollar);
+    lengthVar.storage_class |= STC.static_ | STC.const_;
+}
+
+/* Same as above, but determines the length from 'type'. */
+package void setLengthVarIfKnown(VarDeclaration lengthVar, Type type)
+{
+    if (!lengthVar)
+        return;
+    if (lengthVar._init && !lengthVar._init.isVoidInitializer())
+        return; // we have previously calculated the length
```

Comment:
Apparently this caching doesn't work.

Reply:
You mean coverage wise? That's probably because semantic only calls it once. Doesn't make the safeguards any less correct.